### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by filippD**
added tests

### CODE REVIEW

File: spec/board_spec.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
+require './lib/board' 
+
+RSpec.describe Board do
+  describe "initialize" do
+    it "returns an array with nil cells value" do
+      board = Board.new(3)
+      expect(board.matrix).to eql([nil,nil,nil,nil,nil,nil,nil,nil,nil])
```
Comment (heyapricot) [inline spec/board_spec.rb:7]: Why do you need to test this? As a rule of thumb, test the `public` methods. The way to test is: Am I receiving what I expected to receive?

File: spec/board_spec.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
+require './lib/board' 
+
+RSpec.describe Board do
+  describe "initialize" do
```
Comment (heyapricot) [inline spec/board_spec.rb:4]: Many time you'll want to use the same object for multiple test case. For that use [`let or let!`](http://www.betterspecs.org/#let)

File: spec/board_spec.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    it "returns an array with nil cells value" do
+      board = Board.new(3)
+      expect(board.matrix).to eql([nil,nil,nil,nil,nil,nil,nil,nil,nil])
+    end
+    it "returns an array with combinaisons win value" do
```
Comment (heyapricot) [inline spec/board_spec.rb:9]: Typo: `combinations`

File: spec/board_spec.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      board = Board.new(3)
+      expect(board.win_combinations).to eql([[0, 1, 2],[3, 4, 5],[6, 7, 8],[0, 3, 6],[1, 4, 7],
+      	[2, 5, 8],[0, 4, 8],[2, 4, 6]])
```
Comment (heyapricot) [inline spec/board_spec.rb:12]: Again, I don't think you need to test for this. This is more related to the inner workings of your class

File: spec/board_spec.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      board.matrix = ["x","x","o",nil,nil,nil,nil,nil,nil]
+      expect(board.full?).to eql(false)
+    end
+
+    it "returns true if there is no cells = nil" do
+      board = Board.new(3)
+      board.matrix = ["x","x","o","o","x","o","x","o","x"]
+      expect(board.full?).to eql(true)
+    end  
+  end
```
Comment (heyapricot) [inline spec/board_spec.rb:30]: Notice how you can directly assign a value to `board.matrix` ? That is risky. I could do `board.matrix = 'DANGER'` and the program would take it. It's your choice to improve this. You can make a method that receives an array, validates it and if it's valid it assigns @matrix a copy of the array received by the method.

File: spec/board_spec.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+  describe "player_won" do
+  	it "return true if the player puts his symbol in all the cells of one of win combinaisons" do
+      board = Board.new(3)
+  		board.matrix = ["x","x","x","o","o","nil","nil","nil","nil"]
+  		expect(board.player_won?("x")).to eql(true)
+  	end	
+  end
```
Comment (heyapricot) [inline spec/board_spec.rb:38]: You are only testing 1 scenario, the board has it's first row full of `x`. What happens when:
- Other rows are full?
- What about the columns?
- The Diagonals?
- What if the board is full and there is not a sequence?

File: spec/board_spec.rb (hunk 2 above)
Comment (heyapricot) [reply spec/board_spec.rb:4]: Hint: you can use let at least for an instance of `Board`

File: spec/game_spec.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    it "returns false if sumbol is longer than single character" do
+    	expect(game.valid_symbol?('ab')).to be false 
+    end
+    it "returns false if symbol is taken by another player" do
+    	expect(game.valid_symbol?('X')).to be false
+    end
+    it 'returns true if all is good' do 
+    	expect(game.valid_symbol?('O')).to be true
+  	end
+  end
```
Comment (heyapricot) [inline spec/game_spec.rb:19]: Should `valid_symbol?` be a public method?

File: spec/game_spec.rb [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  	it "returns false if player asks input is higher than 9" do
+    	expect(game.valid_board_size?(10)).to be false 
+    end
+    it "returns false if player asks input is less than 2" do
+    	expect(game.valid_board_size?(1)).to be false 
+    end
+	it 'returns true if all is good' do 
+    	expect(game.valid_board_size?(5)).to be true
+  	end
+  end
```
Comment (heyapricot) [inline spec/game_spec.rb:32]: Should `valid_board_size?` be a public method?

File: spec/game_spec.rb [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+
+  describe "want_to_quit?" do
+    it "should return true if length of the array is one and his value is q" do
+      expect(game.want_to_quit?(['q'])).to be true
+    end  
+    it "should return false if length of the array is great than one and his value is not q" do
+      expect(game.want_to_quit?([3,'B'])).to be false
+    end  
+  end
```
Comment (heyapricot) [inline spec/game_spec.rb:42]: Not sure if this should be a public method

File: spec/game_spec.rb [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+  end 
+
+  describe "move_index" do
+    it "should return an integer" do
+      expect(game.move_index(['1','b'])).to be_a(Integer)
+    end  
+    it "should return th right array index" do
+      expect(game.move_index(['0','a'])).to eql(0)
+    end  
+  end
```
Comment (heyapricot) [inline spec/game_spec.rb:51]: Probably a private method

File: spec/game_spec.rb [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    end  
+
+    it "should return false if input is out of board range" do
+      expect(game.valid_input?(['99','A'])).to be false
+      expect(game.valid_input?(['1','Z'])).to be false
+    end  
+    it "should return true if all is good" do
+      expect(game.valid_input?(['1','B'])).to be true
+    end 
+  end
```
Comment (heyapricot) [inline spec/game_spec.rb:65]: Probably private method

File: spec/game_spec.rb [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+  describe "empty_space?" do
+    it "should return false if matrix[i] is not nill" do
+      game.board.matrix[0] = 'X'
+      expect(game.empty_space?(0)).to be false
+    end
```
Comment (heyapricot) [inline spec/game_spec.rb:71]: Probably private method

File: spec/game_spec.rb [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    end 
+  end
+
+  describe "empty_space?" do
+    it "should return false if matrix[i] is not nill" do
+      game.board.matrix[0] = 'X'
+      expect(game.empty_space?(0)).to be false
+    end 
+
+    it "should return true if matrix[i] is nill" do
```
Comment (heyapricot) [inline spec/game_spec.rb:73]: Typo: `nil`

File: spec/player_spec.rb [hunk 14]
```diff
@@ -0,0 +1,10 @@
+require './lib/player'
+RSpec.describe Player do
+  describe "initialize" do
+    it "initialize a player" do
+      p = Player.new("zina","x")
+      expect(p.name).to eql("zina")
+      expect(p.symbol).to eql("x")
+    end
+  end
+end
```
Comment (heyapricot) [inline spec/player_spec.rb:10]: There's not much to test here since it has no methods

**Review by heyapricot — CHANGES_REQUESTED** [review]
# Let's use *let* 🤙🏼
👋🏼 Hello! You wrote specs for a big project! Yay! How does that feel?
My observations:
- On board.rb
1. You can use *let* to have a single declaration at the beginning of the describe block ( Using let also provides more advantages)

- General
You need to decide which methods are going to be private and the ones that are going to be public. You tested the validators mostly. Do you really need to expose those methods?
The idea behind the testing is to simulate the way you are going to use your program and your program calls. Here are some ideas for testing scenarios:

- If I perform a write from the `Game` object by using move, is it reflected in the board?
- If the board is almost full and I do the final write, does `Game ` tells me that there was a draw?
- If I change the players in the menu are the names updated?

Think about how you interact with your methods and the expected outcome and test for that. 

Please address my comments and after that submit a new code review request. 
Feel free to reach over the student dashboard for further clarification.

**Review by heyapricot — COMMENTED** [review]
Also: Can we have a specific branch for the development of the tests? Thanks!

File: spec/board_spec.rb (hunk 2 above)
Comment (filippD) [reply spec/board_spec.rb:4]: I honestly could not understand how to use let, all the examples have some passing methods, to increase count for example. I've used before(:each), but I've read it's worse than let. Could you please write an example of declaring board with let?

**Pull request merged**